[v16] Auto-generate operator CRD docs

ptgott opened the pull request:
Backports #44107

* Auto-generate operator CRD docs

We already pull JSON schema data from Teleport protobuf message definitions in order to generate CRD manifests. Extract a shared library from code within the existing protoc plugin for generating CRDs. This way, we can use the library to generate docs for our CRDs with a separate plugin, and don't need to complicate the interface for using the existing plugin.

- Organize `integrations/crdgen` into `lib` and `cmd` directories, turning `lib` into a package.
- Edit the `generateSchema` signature to include a function that formats a CRD into a document, injecting separate functions for formatting the CRD as YAML and as a docs page.
- Add Makefile targets for building the docs plugin and generating CRD docs.

To display the resulting reference pages in the docs, this change configures `docs/config.json` to generate the Reference section of the sidebar from the docs directory. This requires moving the Helm reference includes directory into `docs/pages/includes`.

* Respond to hugoShaka feedback

- Do not display the `status` field.
- Indicate that enum fields can take an integer or a string. This is consistent with our comment conventions, which describe both the integer and the string values for each enum option.

* Sort CRD reference page sections

Also handle duplicate page sections by including the names of parent properties within section headings.

* Fix crdgen Makefile targets

- bring test data and hack out of `lib/`
- fix broken debug builds
- add crd-docs checks to the `crds-up-to-date` target
- trigger proto lint on CRD and CRD doc changes
- removed stale gitkeep in `docs/pages/includes/crd-docs/.gitkeep`
- split crdgen formatting logic in another file for readability
- update license header years on new files

* fixup! Fix crdgen Makefile targets

* lint files due to false positive 'generated' comment

* fixup! lint files due to false positive 'generated' comment

* Incorporate rosstimothy feedback

Remove an unnecessary package.

Also regenerate the CRD docs.

* fix debug builds and gitignore

* fixup! fix debug builds and gitignore

* Respond to rosstimothy feedback

Fix integrations/operator/crdgen/Makefile

* Fix spelling

* Fix flaky propertyTable test

Compare the unsorted slice elements, since sorting takes place on the containing VersionSection.

---------

ptgott committed:
Auto-generate operator CRD docs

Backports #44107

* Auto-generate operator CRD docs

We already pull JSON schema data from Teleport protobuf message
definitions in order to generate CRD manifests. Extract a shared library
from code within the existing protoc plugin for generating CRDs. This
way, we can use the library to generate docs for our CRDs with a
separate plugin, and don't need to complicate the interface for using
the existing plugin.

- Organize `integrations/crdgen` into `lib` and `cmd` directories,
  turning `lib` into a package.
- Edit the `generateSchema` signature to include a function that formats
  a CRD into a document, injecting separate functions for formatting the
  CRD as YAML and as a docs page.
- Add Makefile targets for building the docs plugin and generating CRD
  docs.

To display the resulting reference pages in the docs, this change
configures `docs/config.json` to generate the Reference section of the
sidebar from the docs directory. This requires moving the Helm reference
includes directory into `docs/pages/includes`.

* Respond to hugoShaka feedback

- Do not display the `status` field.
- Indicate that enum fields can take an integer or a string. This is
  consistent with our comment conventions, which describe both the
  integer and the string values for each enum option.

* Sort CRD reference page sections

Also handle duplicate page sections by including the names of parent
properties within section headings.

* Fix crdgen Makefile targets

- bring test data and hack out of `lib/`
- fix broken debug builds
- add crd-docs checks to the `crds-up-to-date` target
- trigger proto lint on CRD and CRD doc changes
- removed stale gitkeep in `docs/pages/includes/crd-docs/.gitkeep`
- split crdgen formatting logic in another file for readability
- update license header years on new files

* fixup! Fix crdgen Makefile targets

* lint files due to false positive 'generated' comment

* fixup! lint files due to false positive 'generated' comment

* Incorporate rosstimothy feedback

Remove an unnecessary package.

Also regenerate the CRD docs.

* fix debug builds and gitignore

* fixup! fix debug builds and gitignore

* Respond to rosstimothy feedback

Fix integrations/operator/crdgen/Makefile

* Fix spelling

* Fix flaky propertyTable test

Compare the unsorted slice elements, since sorting takes place on the
containing VersionSection.

---------

Co-authored-by: hugoShaka <hugo.hervieux@goteleport.com>

hugoShaka commented on docs/cspell.json:
```diff
@@ -838,6 +842,72 @@
+    "specdenyimpersonate",
+    "specdenyjoin",
+    "specdenykubernetes",
+    "specdenyrequest",
+    "specdenyrequestclaims",
+    "specdenyrequestthresholds",
+    "specdenyrequire",
+    "specdenyreview",
+    "specdenyrules",
+    "specdenyspiffe",
+    "specgcp",
+    "specgcpallow",
+    "specgithub",
+    "specgithuballow",
+    "specgitlab",
+    "specgitlaballow",
+    "specgrants",
+    "speckubernetes",
+    "speckubernetesallow",
+    "speckubernetesstatic",
+    "specmappings",
+    "specmembership",
+    "specoidc",
+    "specoptions",
+    "specoptionscert",
+    "specoptionsidp",
+    "specoptionsidpsaml",
+    "specoptionsrecord",
+    "specowner",
+    "specowners",
+    "specownership",
+    "specrotation",
+    "specrotationschedule",
+    "specsaml",
+    "specsigning",
+    "specspacelift",
+    "specspaceliftallow",
+    "specteams",
+    "spectpm",
+    "spectpmallow",
```
This won't scale, can we skip the spell check on the links or those pages instead?

ptgott replied:
I have a PR to address this on `master` here: https://github.com/gravitational/teleport/pull/46259

ptgott committed:
Ignore Operator references in the spellcheck

Backports #46259

Teleport Operator reference docs contain a large number of internal link
references that fail the spell checker. Until we find a more scalable
spell check workflow (#46191), ignore these reference docs in the spell
checker.

ptgott committed:
Update Operator docs

ptgott committed:
Fix lint error
